src/libraries/System.Private.CoreLib/src/System/DefaultBinder.cs

```diff
@@ -196,43 +196,54 @@ internal partial class DefaultBinder : Binder
                     if (pCls.IsByRef)
                         pCls = pCls.GetElementType()!;
 
-                    // the type is the same
-                    if (pCls == argTypes[paramOrder[i][j]])
-                        continue;
+                    int index = paramOrder[i][j];
+                    if (index < args.Length)
+                    {
+                        // the type is the same
+                        if (pCls == argTypes[index])
+                            continue;
 
-                    // a default value is available
-                    if (defaultValueBinding && args[paramOrder[i][j]] == Type.Missing)
-                        continue;
+                        // a default value is available
+                        if (defaultValueBinding && args[index] == Type.Missing)
+                            continue;
 
-                    // the argument was null, so it matches with everything
-                    if (args[paramOrder[i][j]] == null)
-                        continue;
+                        // the argument was null, so it matches with everything
+                        if (args[index] == null)
+                            continue;
 
-                    // the type is Object, so it will match everything
-                    if (pCls == typeof(object))
-                        continue;
+                        // the type is Object, so it will match everything
+                        if (pCls == typeof(object))
+                            continue;
 
-                    // now do a "classic" type check
-                    if (pCls.IsPrimitive)
-                    {
-                        if (argTypes[paramOrder[i][j]] == null || !CanChangePrimitive(args[paramOrder[i][j]]?.GetType(), pCls))
+                        // now do a "classic" type check
+                        if (pCls.IsPrimitive)
                         {
-                            break;
+                            if (argTypes[index] == null || !CanChangePrimitive(args[index]?.GetType(), pCls))
```

Comment:
```suggestion
                            if (argTypes[index] == null || !CanChangePrimitive(args[index].GetType(), pCls))
```

The `args[index]` was checked for `null` above so can be elided here.